brayns/common/scene/Model.h

```diff
@@ -0,0 +1,241 @@
+/* Copyright (c) 2015-2017, EPFL/Blue Brain Project
+ * All rights reserved. Do not distribute without permission.
+ * Responsible Author: Cyrille Favreau <cyrille.favreau@epfl.ch>
+ *
+ * This file is part of Brayns <https://github.com/BlueBrain/Brayns>
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#ifndef Model_H
+#define Model_H
+
+#include <brayns/api.h>
+#include <brayns/common/BaseObject.h>
+#include <brayns/common/geometry/Cone.h>
+#include <brayns/common/geometry/Cylinder.h>
+#include <brayns/common/geometry/Sphere.h>
+#include <brayns/common/geometry/TrianglesMesh.h>
+#include <brayns/common/types.h>
+
+SERIALIZATION_ACCESS(ModelTransformation)
+SERIALIZATION_ACCESS(Model)
+SERIALIZATION_ACCESS(ModelDescriptor)
+
+namespace brayns
+{
+/**
+* @brief The ModelTransformation struct defines the translation, rotation and
+* scaling parameters to be applied to a model
+*/
+struct ModelTransformation : public BaseObject
+{
+    ModelTransformation() = default;
+    ModelTransformation(ModelTransformation&& rhs) = default;
+    ModelTransformation& operator=(ModelTransformation&& rhs) = default;
+
+    Vector3f& getTranslation() { return _translation; }
+    void setTranslation(const Vector3f& value)
+    {
+        _updateValue(_translation, value);
+    }
+    Vector3f& getScale() { return _scale; }
+    void setScale(const Vector3f& value) { _updateValue(_scale, value); }
+    Vector3f& getRotation() { return _rotation; }
+    void setRotation(const Vector3f& value) { _updateValue(_rotation, value); }
+private:
+    Vector3f _translation{0.f, 0.f, 0.f};
+    Vector3f _scale{1.f, 1.f, 1.f};
+    Vector3f _rotation{0.f, 0.f, 0.f};
```

Comment:
I guess my opinion doesn't count very much here, but this looks like Euler angles, which are quite cumbersome to specify formally (is this xzx, xyz, xyx, intrinsic, extrinsic, ...?). The interpretation is not even documented here.
Why don't you use a quaternion or an angle, axis representation, which are unambiguous and self-explanatory?

Reply:
I agree, I guess we will match these with what the user uses in the UI.

Reply:
It's a discussion that comes and go on a regular basis ;) I don't have a strong opinion there, I simply find the translation/scaling/rotation more human readable, and since this is what is exposed to the HTTP interface, I decided to use that representation. I guess the JS guy's opinion also is to consider. I'll ask input from @rolandjitsu

Reply:
This interface doesn't have to match what users see necessarily. This class could use a matrix to encode the transformations and the client could use whatever is preferable depending on the use case, converting it to a matrix at the last step. That would be my choice, in fact.

Reply:
I know this discussion is recurrent and I don't what to start it again here. My criticism is that the way the rotation is expressed in this API is unclear (If you give this to two different people, the chances of ending up with a different transformation are quite high). At the very least, it should be documented better. And since it's not so easy to document easily, I suggest to use something which is unambiguous instead.

Reply:
Any method for manipulating objects from the GUI in a natural way and relative to the camera position will have to do some transformations, no matter what this API is, and for that you need an algebraic representation (matrix or quaternion), which means that in the end you need to transform the user input to a matrix or quaternion.
If this API describes the rotation as something else, it means that you need to transform the algebraic representation to whatever this API requires.  Any representation can be easily converted to a quaternion or a matrix, but I find the opposite more annoying.
For me these are good reasons to choose a quaternion.